File: Neetcode 150/Binary Search/search-a-2d-matrix.cpp

// https://leetcode.com/problems/search-a-2d-matrix/description/

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        
        int rowSize = matrix.size();
        if(rowSize == 0) return false;
        int colSize = matrix[0].size();

        int start = 0, end = rowSize*colSize-1;
        while(start <= end){
            int mid = start + (end-start)/2;
            int row = mid/colSize;
            int col = mid%colSize;
            
            if(matrix[row][col] == target) return true;
            else if(matrix[row][col] > target) end = mid-1;
            else start = mid+1;
        }

        return false;
    }
};

File: Neetcode-150-js/Binary Search/search-a-2d-matrix.js

// https://leetcode.com/problems/search-a-2d-matrix/description/

/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    let rowSize = matrix.length;
    if(rowSize === 0) return false;
    let colSize = matrix[0].length;
    
    let start = 0, end = rowSize*colSize-1;
    while(start <= end){
        let mid = Math.floor((start+end)/2);
        let row = Math.floor(mid/colSize);
        let col = mid%colSize;

        if(matrix[row][col] === target) return true;
        else if(matrix[row][col] < target) start = mid+1;
        else end = mid-1;
    }

    return false;
};